feat: allow setting fields as not nullable

akhilnarang opened the pull request:
Resolves #21824

This adds in a `not nullable` checkbox for docfields
If selected - this will ensure that a sane default value (0 / "") will get set instead of `NULL`

akhilnarang committed:
feat(docfield): add in a nullable checkbox field

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: make return type annotation make more sense

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: set a non-null value if docfield isn't set as nullable

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: drop unused variables and parameters

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: check for docfield not_nullable property to decide whether a field can be nullable

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(DbColumn): set `non_nullable` attribute and make use of it for column definition

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: allow ignoring implicit commit warning

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: migrate columns to be non-nullable if required

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/core/doctype/docfield/docfield.json:
```diff
@@ -559,13 +560,19 @@
+  {
+   "default": "0",
+   "fieldname": "not_nullable",
+   "fieldtype": "Check",
```
Hide it in UI when it doesn't make sense:
- Fieldtypes that are of "no value" types
- Fieldtypes that always have defaults in framework - numeric types.

ankush commented on frappe/database/mariadb/database.py:
```diff
@@ -310,7 +310,7 @@ def create_user_settings_table(self):
 		)
 
 	@staticmethod
-	def get_on_duplicate_update(key=None):
```
This is required I think, because interface requires it. I am guessing it was added for postgres. 

https://github.com/frappe/frappe/blob/567cfda0e309cd3ed665b7ddc9ce8ca551f15e0f/frappe/database/postgres/database.py#L330-L333

akhilnarang replied:
Ah makes sense

akhilnarang replied:
Nvm, IDE had pointed it out to me, so I double checked, the method isn't defined in the superclass, so that's why it didn't matter.

Also seems like its not called anywhere anyway :rofl: 

![image](https://github.com/frappe/frappe/assets/10119037/d8800de6-f4a5-457b-97fd-08ade395c810)

ankush commented:
@akhilnarang we follow same rule for PR title as commits: https://www.conventionalcommits.org/en/v1.0.0/

akhilnarang commented:
> @akhilnarang we follow same rule for PR title as commits: https://www.conventionalcommits.org/en/v1.0.0/

Ah yeah, I had added for most PRs, missed it here :man_facepalming:

akhilnarang committed:
chore(mariadb): switch to using query builder for update statement

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: rework implicit commit checks

Add in some more "reset" queries
- begin
- start

Also check *after* resetting

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: don't check for semgrep here

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: simplify transaction handling

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: move check_implicit_commit() call back to its original location

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/database/mariadb/schema.py:
```diff
@@ -102,11 +102,23 @@ def alter(self):
+			if col.not_nullable:
+				try:
+					table = frappe.qb.DocType(self.doctype)
+					frappe.qb.update(table).set(col.fieldname, "").where(
```
This should be default and not empty string?

ankush commented on frappe/database/schema.py:
```diff
@@ -175,7 +177,17 @@ def alter(self):
+		options,
+		unique,
+		precision,
+		not_nullable,
```
Lets force these to be kwargs. This order has no meaning.

ankush commented on frappe/database/schema.py:
```diff
@@ -208,6 +221,9 @@ def get_definition(self, for_modification=False):
 			column_def += f" default {frappe.db.escape(self.default)}"
 
+		elif self.not_nullable:
+			column_def += f" not null default {frappe.db.escape('')}"
```
Umm, default is already set on line above.

akhilnarang replied:
Hmm, I guess I should split out this from the above if-elif chain and keep it standalone like `unique`.

ankush commented on frappe/model/base_document.py:
```diff
@@ -384,6 +384,10 @@ def get_valid_dict(
 
+			# If the docfield is not nullable - set a default non-null value
+			if value is None and getattr(df, "not_nullable", False):
+				value = ""
```
This should be handled much earlier? Also defaults depend on fieldtype.

akhilnarang replied:
A bit confused about this, initially I did set it up to get the default value dynamically - e76c806fed40bc59c288e0937f41be1c00c76cfa

But from what I can understand, its only the varchar types that require this, correct?

ankush commented on frappe/model/document.py:
```diff
@@ -208,7 +208,7 @@ def get_latest(self):
 
 		return self._doc_before_save
 
-	def check_permission(self, permtype="read", permlevel=None):
```
I know this is useless but there's probably a code that sends this and will break. I remember removing such code just few weeks ago.

ankush reviewed (changes requested):
Add tests for all the use cases:
- new doctype with nullable set and effect on schema
- existing doctype that changes from nullable to non-nullable and vise versa. Create some data in the test doctype before migrating so we can check the correctness end to end.

akhilnarang committed:
chore(DbColumn): force constructor arguments to be kwargs

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: add in tests for non-nullable field

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: cleanup `DbColumn.get_definition()`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: make use of `non_nullable` to avoid ifnull usage

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/database/database.py:
```diff
@@ -960,16 +963,24 @@ def begin(self, *, read_only=False):
-	def commit(self):
+	def commit(self, new_transaction: bool = True):
 		"""Commit current transaction. Calls SQL `COMMIT`."""
+		if not self.in_transaction:
+			self.logger.warn("Not in transaction, ignoring commit")
+			return
```
IMO we should never change behaviour of db.commit because even if we haven't made writes or in transaction, there are valid usage of committing/rolling back.

akhilnarang replied:
Reverted this, switched to `frappe.db.sql_ddl()` where required to avoid implicit commit issue.

ankush commented on frappe/core/doctype/doctype/doctype.py:
```diff
@@ -364,16 +364,23 @@ def setup_fields_to_fetch(self):
+						if df.not_nullable:
+							update_query += "WHERE `target`.`{fieldname}`=''"
+						else:
+							update_query += "WHERE ifnull(`target`.`{fieldname}`, '')=" ""
```
```suggestion
							update_query += "WHERE ifnull(`target`.`{fieldname}`, '')='' "
```

ankush commented on frappe/database/mariadb/schema.py:
```diff
@@ -102,11 +103,24 @@ def alter(self):
+				try:
+					table = frappe.qb.DocType(self.doctype)
+					frappe.qb.update(table).set(col.fieldname, default_value).where(
+						getattr(table, col.fieldname).isnull()
```
```suggestion
						table[col.fieldname].isnull()
```

ankush commented on frappe/database/mariadb/schema.py:
```diff
@@ -102,11 +103,24 @@ def alter(self):
+					frappe.qb.update(table).set(col.fieldname, default_value).where(
+						getattr(table, col.fieldname).isnull()
+					).run()
+				except Exception as e:
```
Brainstorm what are the failure modes here? 
- Db constraints causing it to fail - don't allow invalid db constraints.

ankush commented on frappe/model/base_document.py:
```diff
@@ -384,6 +385,10 @@ def get_valid_dict(
 				continue
 
+			# If the docfield is not nullable - set a default non-null value
+			if value is None and getattr(df, "not_nullable", False):
+				value = get_not_null_defaults(df.fieldtype)
```
What is doctype has non-empty string default? 


end goal: `frappe.new_doc()` should always have the defaults set.

ankush commented on frappe/model/db_query.py:
```diff
@@ -109,7 +109,6 @@ def execute(
 		save_user_settings=False,
 		save_user_settings_fields=False,
 		update=None,
-		add_total_row=None,
```
Force kwargs using `, *, ` where it makes sense.

ankush commented on frappe/model/db_query.py:
```diff
@@ -808,7 +807,10 @@ def prepare_filter_condition(self, f):
+			if df and (
+				df.fieldtype in ("Check", "Float", "Int", "Currency", "Percent")
+				or getattr(df, "not_nullable", False)
+			):
```
Implement this for `in` and `not in` operator too and for NestedSet operators too.

ankush replied:
Also check if `can_be_null` is again being set by some other heuristics down below (e.g. `not set`)

akhilnarang committed:
fix(doctype): update query was broken

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: use index to access table column instead of getattr

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: go back to original transaction handling mechanism

This reverts the following:
07acfeed47 chore: move check_implicit_commit() call back to its original location
fc38a0b503 chore: simplify transaction handling
1e29e81543 chore: rework implicit commit checks
ae0a3fd202 feat: allow ignoring implicit commit warning

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented:
- [ ] lets make it clear that this is only for varchar/string types. 
- [ ] some `string types` don't support setting defaults in strict mode. Worth checking out.

akhilnarang committed:
refactor(mariadb/schema): use sql_ddl() method

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(exporter): handle docfield being non-nullable

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat: use user-specified default value if passed

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(db_query): check for docfield not_nullable

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(mariadb/schema): use correct default value

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
feat(postgres/schema): add support for setting fields as not-nullable

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(postgres): properly change nullability and default values

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: use default default if a custom default isn't specified

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: cleanup column definition

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: adapt tests to actually work

Couldn't seem to sanely get postgres default state

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented:
@akhilnarang can you resolve the conflicts and fix tests. Seems mostly ready for merge, will do a final review after the tests pass.

akhilnarang committed:
fix: 0 is falsy, set correct defaults

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: get correct column definition

Signed-off-by: Akhil Narang <me@akhilnarang.dev>
